### PR TITLE
Update challenges.tex

### PR DESCRIPTION
**Pull request by AGEdwards**

### CODE REVIEW

File: final_report/Chapter5-Implementation/challenges.tex [hunk 1]
```diff
@@ -0,0 +1,27 @@
+\subsubsection{Overall Functionality}
+\par
+The \textit{Challenges} microservice was created to set personal and group goals. When creating a challenge, a member has the ability to select an activity, goal metric and goal within a time frame. This microservice is the only one to use the coordinator user type, with coordinators being able to create group challenges. The group challenges are specific to a single group, and can only be joined by users within that group.
+
+\subsubsection{MVC Architecture}
+\par
+As the microservice requires CRUD operations on data, it was essential that a persistence layer was present.
+Initially the microservice consisted of \textit{Challenge} and \textit{Activity} models, but this proved to be inadequate when it came to database normalisation. After many alterations to the architecture, \textit{UserChallenge}, \textit{ChallengeManage}, \textit{Activity} and \textit{GoalMetric} were the final controllers, with \textit{ChallengeManage} only named as such due to an issue with \textit{nginx}. The \textit{UserChallengeController} is used to keep track of challenges a user is currently taking part in, which includes both personal and group challenges.
```
Comment (dkmonaghan) [inline final_report/Chapter5-Implementation/challenges.tex]: Not really sure what this means. You talk about the models and say they're inadequate, then talk about alternations to the controllers?

File: final_report/Chapter5-Implementation/challenges.tex [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+\par
+Each user challenge is assigned a \textit{ChallengeId}. The view page for \textit{UserChallenge} displays the challenge information for the current user's challenges. The \textit{ChallengeManageController} is used to handle all challenges that have been created. The view page for \textit{ChallengeManage} displays all group challenges. A challenge is assigned both a \textit{GoalMetricId} and \textit{ActivityId}. The \textit{ActivityController} and \textit{GoalMetricController} are used to allow activities and goal metrics to be changed by administrators and coordinators. The activities are retrieved from the \textit{Health Data Repository} and the goal metrics need to be manually entered to match the database key.
```
Comment (dkmonaghan) [inline final_report/Chapter5-Implementation/challenges.tex]: This paragraph is pretty not fun to read with every sentence starting with 'The' and so on. Can this be revised?

File: final_report/Chapter5-Implementation/challenges.tex [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+\subsubsection{API Access}
+\par
+The microservice requires access to the \textit{Health Data Repository} microservice to retrieve data in order to update the progress of a challenge. When the user challenges page or API endpoint is accessed, their challenges are updated with the most recent data.
+The microservice also requires access to the \textit{User Groups} microservice to retrieve information about a user or group. When a user loads the group challenges page, the challenges are filtered depending on to which group they belong.
+The microservice also requires access to the \textit{Communications} microservice, in order to send notifications to users whenever their challenges close.
```
Comment (arglanc) [inline final_report/Chapter5-Implementation/challenges.tex]: Slightly repetitive here with "The microservice" 3 times.

Comment (dkmonaghan) [reply final_report/Chapter5-Implementation/challenges.tex]: "slightly"

**Pull request merged**